### PR TITLE
feat: add -H flag to use hard links instead of symlinks

### PR DESCRIPTION
**Pull request by mbhall88**
Adds the `-H` option to use hard links (`ln`) instead of symbolic links (`ln -s`) when updating a repository and linking existing files to a new datestamped directory. 

This prevents inode exhaustion on HPC systems when updating large databases, as hard links point to the existing data and do not consume additional inodes. It also allows for seamless copying to compute nodes without needing to dereference symlinks.

A bit of background: I've run into a significant bottleneck when running `genome_updater.sh` a second time to check for missing or updated files on our institutional HPC cluster. When `genome_updater` creates a new datestamped directory for the updated run, it leaves the previously downloaded files in their original folder and creates symbolic links for every file it already had. In my recent run to grab 14,341 new/fixed files, the tool generated 3,414,730 symbolic links.

On some HPC systems, storage quotas are often dictated by number of files (inodes), as well as file size. Because every symbolic link requires a brand new inode, generating 3.4 million symlinks instantly triggered our group's inode limits.

I then ran into a secondary issue: When shovelling all these downloads over to compute-node local storage ($TMPDIR) for processing, standard copy commands just copy the symlinks rather than resolving them. I only realised this when my downstream pipeline failed because the millions of actual files were still sitting in the old directory, not the current datestamped folder.

I hope this feature is in scope and that my implementation is acceptable? Happy to make any other changes you would like. And, of course, feel free to decline the update if you don't want it.

### CODE REVIEW

**Comment by mbhall88** [comment]
I guess the main disadvantage of the hard links is that you can't hard link files across filesystems. Though I suspect it is extremely rare that people are going to be updating across filesystems?  
And a small minor thing is that when doing `ls -l` the user won't see the nice pointer to the location of the "original" file.

If you want hard links to be the default I can update the PR so that the new options is the inverse of my suggestion? i.e., `-H` will use soft links and the default is hard links?

**Comment by pirovc** [comment]
I will perform some tests and check some edge cases before I give you a go. Thanks for volunteering :)

**Comment by pirovc** [comment]
All good from my side, you can go on with the implementation. Thanks again. Just one request: instead of a boolean, please implement as a string parameter `-H hard`/`-H soft`. That way the parameter is stored between updates and also enable the addition of further options in the future (e.g. `copy`, `move`).

Remember to rebase your branch, I had to do a hotfix in between.

**Comment by mbhall88** [comment]
Okay, this should all be done as requested now.

File: genome_updater.sh [hunk 1]
```diff
@@ -759,8 +763,11 @@ remove_files() # parameter: ${1} file, ${2} fields [assembly_accesion,url] OR fi
     deleted_files=0
     while read -r f; do
         path_name="${target_output_prefix}$(path_output "${f}")${f}"
+        # check hard link count.
+        hl_count=$(stat -c '%h' "${path_name}" 2>/dev/null || stat -f '%l' "${path_name}" 2>/dev/null || echo 1)
```
Comment (pirovc) [inline genome_updater.sh:767]: why is the double validation needed here?

Comment (mbhall88) [reply genome_updater.sh:767]: It's essentially just a portability "hack". In GNU `stat`, the way to retrieve the hard link count is `-c '%h'`. But in BSD (MacOS) `stat`, `-c` is not present and is instead `-f` and the "format" code is `%l`. Super annoying when there is these differences between GNU and BSD utils. 

If you want I can wrap this in a function somewhere or try harder for a portable solution?

File: genome_updater.sh [hunk 2]
```diff
@@ -759,8 +763,11 @@ remove_files() # parameter: ${1} file, ${2} fields [assembly_accesion,url] OR fi
     deleted_files=0
     while read -r f; do
         path_name="${target_output_prefix}$(path_output "${f}")${f}"
+        # check hard link count.
```
Comment (pirovc) [inline genome_updater.sh]:
```suggestion
        # check hard link count: GNU stat: -c '%h' / BSD stat: -f '%l'
```

**Comment by pirovc** [comment]
Thanks, again. Merging on dev, need to update some other stuff and make a release soon.
